js/private/translate_pnpm_lock.bzl

```diff
@@ -0,0 +1,364 @@
+        integrity = resolution.get("integrity")
+        if not integrity:
+            msg = "package %s resolution has no itegrity field" % packagePath
+            fail(msg)
+        dev = resolution.get("dev", False)
+        optional = resolution.get("optional", False)
+        has_bin = resolution.get("hasBin", False)
+        requires_build = resolution.get("requiresBuild", False)
+        package_info = {
+            "name": package,
+            "version": version,
+            "integrity": integrity,
+            "dependencies": {},
+            "dev": dev,
+            "optional": optional,
+            "has_bin": has_bin,
+            "requires_build": requires_build,
+        }
+        dependencies = []
+        package_deps = packageSnapshot.get("dependencies")
+        if package_deps:
+            for (dep_name, dep_version) in package_deps.items():
+                normalized_version = npm_utils.normalize_version(dep_version)
+                dependencies.append(npm_utils.versioned_name(dep_name, normalized_version))
+        if dependencies:
+            package_info["dependencies"] = dependencies
+        packages[npm_utils.versioned_name(package, version)] = package_info
+    return {
+        "dependencies": direct_dependencies,
+        "packages": packages,
+    }
+
+_NPM_IMPORT_TMPL = \
+    """    npm_import(
+        name = "{name}",
+        integrity = "{integrity}",
+        link_package_guard = "{link_package_guard}",
+        package = "{package}",
+        version = "{version}",{maybe_deps}{maybe_indirect}{maybe_patches}{maybe_patch_args}
+    )
+"""
+
+_ALIAS_TMPL = \
+    """load("//:package.bzl", "package", "package_dir")
+
+alias(
+    name = "{basename}",
+    actual = package("{name}"),
+    visibility = ["//visibility:public"],
+)
+
+alias(
+    name = "dir",
+    actual = package_dir("{name}"),
+    visibility = ["//visibility:public"],
+)"""
+
+_PACKAGE_TMPL = \
+    """
+load("@aspect_rules_js//js/private:npm_utils.bzl", "npm_utils")
+
+def package(name):
+    return Label("@{workspace}//{link_package}:{namespace}__" + npm_utils.alias_target_name(name))
+
+def package_dir(name):
+    return Label("@{workspace}//{link_package}:{namespace}__" + npm_utils.alias_target_name(name) + "__dir")
+"""
+
+def _impl(rctx):
+    if rctx.attr.prod and rctx.attr.dev:
+        fail("prod and dev attributes cannot both be set to true")
+
+    lockfile = _process_lockfile(
+        lockfile = json.decode(rctx.read(rctx.attr.pnpm_lock)),
+        prod = rctx.attr.prod,
+        dev = rctx.attr.dev,
+        no_optional = rctx.attr.no_optional,
+    )
+
+    link_package = rctx.attr.package
+    if link_package == ".":
+        link_package = rctx.attr.pnpm_lock.package
+
+    direct_dependencies = lockfile.get("dependencies")
+    packages = lockfile.get("packages")
+
+    repositories_bzl = [
+        """load("@aspect_rules_js//js:npm_import.bzl", "npm_import")""",
+        "",
+        "def npm_repositories():",
```

Comment:
needs `# buildifier: disable=function-docstring` or else actually add a docstring

Otherwise I get yelled at when I vendor the resulting generated starlark file